nemo_aligner/utils/trt_llm.py

```diff
@@ -0,0 +1,175 @@
+                use_refit=True,
+                reshard_model=self.reshard_model,
+            )
+            self._trtllm_model_compiled = True
+        else:
+            self.trt_llm_exporter.refit(model, self.model_cfg)
+
+    def generate(self, inputs):
+        prompt_tokens, prompt_lengths = inputs
+
+        batch_input_ids = []
+        for idx in range(prompt_tokens.shape[0]):
+            batch_input_ids.append(prompt_tokens[idx][0 : prompt_lengths[idx]].cpu())
+
+        random_seeds = torch.randint(
+            0, 2 ** 32, size=(prompt_tokens.shape[0],), dtype=torch.long, generator=self.rng_generator
+        )
+        self.sampling_config.update(random_seed=random_seeds)
+
+        output_dict = tensorrt_llm_run.tensorrt_llm_worker_context.decoder.generate(
+            batch_input_ids=batch_input_ids, sampling_config=self.sampling_config, streaming=False
+        )
+
+        # remove beam dim from output_ids: [mbs, beam_dim, sequence len]
+        output_ids = torch.squeeze(output_dict["output_ids"], dim=1).long()
+        response_lengths = torch.squeeze(output_dict["sequence_lengths"], dim=1).long()
+        max_length = response_lengths.max().item()
+
+        # TRTLLM with PP erroneously inserts padding so have to remove it here
+        # TRTLLM only produces valid outputs on the source rank, so we can only process it here
+        # and rely on the aligner broadcast to get it to the other ranks. Miraculously, the length
+        # is still correct on the non src ranks
+        if (
+            parallel_state.get_pipeline_model_parallel_world_size() > 1
+            and parallel_state.get_model_parallel_src_rank() == torch.distributed.get_rank()
+        ):
+            valid_tokens = output_ids != self.pad_id
+            # we can't just natively use the response length here
+            # because there are cases where the model generates
+            # stop strings after it has stopped. so we need to
+            # be slightly inefficient and then remove the excess later on
+            valid_token_lengths = valid_tokens.sum(-1, keepdims=True)
+            max_unpadded_length = valid_token_lengths.max()
+
+            _output_ids = torch.full(
+                (response_lengths.size(0), max_unpadded_length),
+                fill_value=self.pad_id,
+                dtype=output_ids.dtype,
+                device=output_ids.device,
+            )
+
+            # only fill up to the amount of valid tokens
+            src_index_mask = (
+                torch.arange(max_unpadded_length, device=response_lengths.device).view(1, -1) < valid_token_lengths
+            )
+
+            _output_ids[src_index_mask] = output_ids[valid_tokens]
+
+            invalid_response_mask = torch.arange(max_unpadded_length, device=response_lengths.device).view(
+                1, -1
+            ) >= response_lengths.view(-1, 1)
+            _output_ids[invalid_response_mask] = self.pad_id
+
+            output_ids = _output_ids
```

Comment:
The whole logic to compute `_output_ids` seems pretty convoluted. I feel like we can probably simplify it, but I don't fully understand the issue with just using `response_lengths`. Can you provide an example of when `response_lengths` is not enough?

Reply:
when we enable TRT-LLM with PP it returns the wrong buffer(known bug in v10).

when we give the buffer to TRT-LLM it's in the following format:
```
[prompt tok, prompt tok, PAD, PAD]
[prompt tok, prompt tok, prompt tok, prompt tok]
```

after generate(only when it's PP) it's
```
[prompt tok, prompt tok, PAD, PAD, resp_tok, resp_tok]
[prompt tok, prompt tok, prompt tok, prompt tok, resp_tok, resp_tok]
```
so I'm not able to directly use the response length, since the response length because taking it up to the response length will include the PAD tokens in the middle. so this logic removes the PAD token

Reply:
I see, thanks. Would you mind adding this example to the comment above the `if`? I think it makes things much clearer to explain what `TRTLLM with PP erroneously inserts padding so have to remove it here` means, and why we need the code below.

Also, I'm wondering what happens if `eos_id` is a stop word: does TRT-LLM count it in `response_lengths` if that's the token that the model generates to end the generation? I think it's important that it does (because we need to compute policy updates on this token), but then we might incorrectly truncate the output by one here and I suspect we should do something like `max_unpadded_length = max(valid_token_lengths.max(), max_length)` in order to be safe.

Reply:
> I see, thanks. Would you mind adding this example to the comment above the if? I think it makes things much clearer to explain what TRTLLM with PP erroneously inserts padding so have to remove it here means, and why we need the code below.

sure! fa9ded94796110896f0af32715e3855d199d91aa

>Also, I'm wondering what happens if eos_id is a stop word: does TRT-LLM count it in response_lengths if that's the token that the model generates to end the generation? I think it's important that it does (because we need to compute policy updates on this token), but then we might incorrectly truncate the output by one here and I suspect we should do something like max_unpadded_length = max(valid_token_lengths.max(), max_length) in order to be safe.

just checked with Daniel, we're pretty sure it does count it(otherwise we'd see more cases where the model doesn't stop properly). In the PPO logging code for text on nemotron(extra id is 1 token) and llama(extra id is not 1 token) I am able to see the `<extra_id_1>` stop token fully, so I think the response length counts it. 

But this actually raises a problem when stop_id == self.pad_id, then I think we're off by 1. I've added the max [here](5f6fcc716da0066d929e984485260847d7235ca0), I will test this along with the non convergence changes. Keeping this thread unresolved until I fully test it

Reply:
Ok thanks -- I'm not worried about `<extra_id_1>` because it's not `tokenizer.eos_id`. Can you maybe just try doing a TRT-LLM generation with
```
            end_id=extra_id_1,
            pad_id=extra_id_1,
```
where `extra_id_1` is the special token associated to `"<extra_id_1>"`?

Then check if `response_length` includes the `<extra_id_1>` token generated to end the response.

Reply:
observations:
1. when (end_id == pad_id) != extra_id_1 -> TRT-LLM includes the "<extra_id_1>" in the response length
2. when end_id == extra_id_1, but pad_id is something else -> TRT-LLM does **not** include the "<extra_id_1>" in the response length
3. when pad_id == extra_id_1, end_id == something else -> TRT-LLM includes the "<extra_id_1>"
4. when end_id==extra_id_1 == pad_id _> TRT-LLM does **not** includes "<extra_id_1>" in the response length

so it looks like for this to work, end_id cannot be an actual end token and we have to use the stop list to end.

Reply:
i'm not sure how to fix this generally.. because it depends on the tokenizer(if the tokenizer.pad_id != tokenizer.end_id) and the model not stopping on end_id. this is generally hard to detect.

maybe we should just raise a very annoying warning or put this in known issues?

Reply:
Thanks a lot for the tests!

It seems that what's going on is what I worried about, i.e., when the response ends on `end_id` then it's not included in the response length (while `pad_id` doesn't really matter -- so I'm not sure why you said it depends on if `tokenizer.pad_id != tokenizer.end_id`, plus anyway we use `self.pad_id = tokenizer.eos_id` and not `tokenizer.pad_id`).

Just to be 100% sure, you still had `"<extra_id_1>"` in `end_strings` in those experiments, right?

I actually believe it's fixable but potentially a bit annoying. What you could do is something like this (sketch):
```python
for generation in all_generations:
    if generation.length < max_generation_length:
        for stop_tokens in stop_list:
            if generation.endswith(stop_tokens):
                break  # ended on one of the stop words
        else:
            # Generation did not end with a stop word and is not max length => must have ended with eos_id
            assert next_token_after_generation_end == eos_id
            generation.length += 1
```

Reply:
> Just to be 100% sure, you still had "<extra_id_1>" in end_strings in those experiments, right?

Yes, and it was stopping on this string.

> so I'm not sure why you said it depends on if tokenizer.pad_id != tokenizer.end_id

Actually, I was thinking about just detokenizing the tokenizer.eos_id and then putting that as if it were an end string -> this way TRT-LLM can use the same logic and we don't need the above. It requires the pad_id != eos_id though. Do you think that's reasonable?

Reply:
> I was thinking about just detokenizing the tokenizer.eos_id and then putting that as if it were an end string

Why would this work? (you confirmed that you had `<extra_id_1>` in `end_strings`, and yet it didn't count this token when `eos_id` was `<extra_id_1>`)

> It requires the pad_id != eos_id in the tokenizer though

Can you explain why?

Reply:
i was thinking of something like this: a50ab3721924cf9c09bc7188042132c72e65f0be.

essentially the trick i think is to just not pass eos_id or anything possible eos_id to trt-llm but add it manually in the stop_list. which means it'll require pad_id(because that's what we pass) to not be the same as eos_id

Reply:
this is a little bit hard to test, but i put end_id == token for "A" and it stopped and included it

Reply:
This should work, thanks!